### ISSUE TITLE
TODO: Webm support

### ISSUE THREAD

**Issue by Apogate** [spootTheLousy/saguaro#86]
We gotta fit in with the cool kids eventually

**Comment by RePod** [comment]
~~The problem with this is GD doesn't support generating thumbnails from videos (WebM).
We'd either have to switch over to something else completely or add a new dependency.~~

No matter what we do, we'd need something able to properly handle WebM (or video in general) to impose restrictions on the uploaded content.

The common method seems to use `ffmpeg` (a **huge** dependency if not compiled from scratch).

If it's already there it would be simple to integrate WebM support (although the jQuery suite would need to be updated for inline expansion and hovering).

This also imposes a huge overhead because we need to manually **queue up** and:
- Check the file size:
  - Can be done safely through PHP and before uploading.
  - Check again after uploading in-case the client-side check failed (or was forged).
- Generate a thumbnail:
  - `ffmpeg -i file.webm -vframes 1 out.jpg`
  - Somehow scale it properly. We can do this with ffmpeg directly or loop it back into my proposed centralized thumbnail creation (which would also have something for ffmpeg).
- If we want to remove audio we need to use a _temporary_ file:
  - `ffmpeg -i file.webm -c:v copy -an temp.webm`
  - Temp file usage may be optional depending on when we determine the filename.

It would also probably be a good idea to centralize thumbnail generation. Currently only _regist/thumb_ generates thumbnails and each instance of obtaining the thumbnail (Catalog, Thread, etc) is its own entity.

---

In short, the general set up is the hard part while guaranteeing cross-platform support. Also on restricted hosts it may be impossible to set up dependencies required for WebM support.

**Comment by malamut2014** [comment]
What about this?<br \>
`shell_exec("ffmpegthumbnailer -s " . max(MAX_W, MAX_H) . " -i $file_location -o $thumb_location") . '!';`

**Comment by RePod** [comment]
- Is this available for Windows? Only seeing Ubuntu (also Debian?).
  - If not, might as well stick with ffmpeg itself.
- If this is used we lose specific advantages of having ffmpeg, such as removing audio.

Might want to look into [containerchan](//github.com/ccd0/containerchan) which has basic dependency-less video thumbnail generation.

**Comment by RePod** [comment]
It may be better to re-implement Regist's thumbnail generation and for video formats attempt to detect ffmpeg (probably in order: relative, path, specified path).

If ffmpeg is available, use it to generate the thumbnail. If not, use a generic thumbnail that indicates a video format.

At that point WebM support would be easy as it'd be as simple as modifying post generation to link to it properly, which could then be handled by the inline expansion and hover features of the jQuery suite.

**Comment by RePod** [comment]
Accepting bug reports specific to #107 as well as contributions for ffmpeg processing (prioritized libav (avconv/avprobe < 0.9)).
